File: 41-50/Q46.py
#46.py
class Solution:
    def permute(self, a):
        res=[]
        a.sort()
        def nextp():
            i=len(a)-2
            while i>=0 and a[i]>=a[i+1]: i-=1
            if i>=0:
                j=len(a)-1
                while a[j]<=a[i]: j-=1
                a[i],a[j]=a[j],a[i]
            a[i+1:]=reversed(a[i+1:])
            return i>=0
        res.append(a[:])
        while nextp():
            res.append(a[:])
        return res

File: 41-50/Q46.cpp
//Q46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& a) {
        vector<vector<int>> res;
        sort(a.begin(), a.end());
        do res.push_back(a);
        while(next_permutation(a.begin(), a.end()));
        return res;
    }
};
